cmd/s3_to_redshift.go

```diff
@@ -0,0 +1,161 @@
+	force           = flag.Bool("force", false, "do we refresh the data even if it's already handled?")
+	dataDate        = flag.String("date", "", "data date we should process, must be full RFC3339")
+	configFile      = flag.String("config", "", "schema & table config to use in YAML format")
+	// things which will would strongly suggest launching as a second worker are env vars
+	// also the secrets ... shhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh
+	host               = os.Getenv("REDSHIFT_HOST")
+	port               = os.Getenv("REDSHIFT_PORT")
+	dbName             = env.MustGet("REDSHIFT_DB")
+	user               = env.MustGet("REDSHIFT_USER")
+	pwd                = env.MustGet("REDSHIFT_PASSWORD")
+	awsRegion          = env.MustGet("AWS_REGION")
+	awsAccessKeyID     = env.MustGet("AWS_ACCESS_KEY_ID")
+	awsSecretAccessKey = env.MustGet("AWS_SECRET_ACCESS_KEY")
+)
+
+func fatalIfErr(err error, msg string) {
+	if err != nil {
+		panic(fmt.Sprintf("%s: %s", msg, err)) // TODO: kayvee
+	}
+}
+
+// in a transaction, truncate, create or update, and then copy from the s3 JSON file
+// yell loudly if there is anything different in the target table compared to config (different distkey, etc)
+func runCopy(db *redshift.Redshift, inputConf s3filepath.S3File, inputTable, targetTable redshift.Table, truncate bool) error {
+	// determine if target table exists
+	// can't compare to empty struct b/c of list
+	createTable := false
+	if targetTable.Name == "" {
+		createTable = true
+	}
+
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+
+	// TRUNCATE for dimension tables, but not fact tables
+	if truncate && !createTable {
+		log.Println("truncating table!")
+		if err = db.RunTruncate(tx, inputConf.Schema, inputTable.Name); err != nil {
+			return fmt.Errorf("err running truncate table: %s", err)
+		}
+	}
+	if createTable {
+		if err = db.RunCreateTable(tx, inputTable); err != nil {
+			return fmt.Errorf("err running create table: %s", err)
+		}
+	} else {
+		if err = db.RunUpdateTable(tx, targetTable, inputTable); err != nil {
+			return fmt.Errorf("err running update table: %s", err)
+		}
+	}
+
+	// COPY direct into it, ok to do since we're in a transaction
+	// assuming that we always want to copy from s3 and have gzip, so last 2 params are true
+	// if we want to change that, we should figure this out from the filename
+	if err = db.RunJSONCopy(tx, inputConf, true, true); err != nil {
+		return fmt.Errorf("err running JSON copy: %s", err)
+	}
+
+	if err = tx.Commit(); err != nil {
+		return fmt.Errorf("err committing transaction: %s", err)
+	}
+	return nil
+}
+
+// This worker finds the latest file in s3 and uploads it to redshift
+// If the destination table does not exist, the worker creates it
+// If the destination table lacks columns, the worker creates those as well
+// The worker also uses a column in the data to figure out whether the s3 data is
+// newer than what already exists.
+func main() {
```

Comment:
Nitpicky, but should main be moved before the lower level functions in this file?